### ISSUE TITLE
Basic infrastructure

### ISSUE THREAD

**Issue by MaxGalindo150** [eigencore/tlama-core#1]
### **Objective**
Create the basic infrastructure of the repository. All modules (directories) containing base code, function implementations, or scripts must reside within the `tlamacore` directory. This directory should include, at a minimum, the following modules:
- `kernels`
- `models`
- `utils`
- `training`
- `data`
- `optim`
- `config`
- `scripts`

---

#### **1. Create the `tlamacore` Directory**
- [x] Initialize the `tlamacore` directory as the root for all core modules.

#### **2. Add Essential Subdirectories**
- [ ] Create the following subdirectories inside `tlamacore`:
  - `kernels`
  - `models`
  - `utils`
  - `training`
  - `data`
  - `optim`
  - `config`
  - `scripts`

#### **3. Define Basic Structure for Each Module**
- [ ] Add `__init__.py` files to each subdirectory to make them importable as Python packages.
- [ ] Define base classes, interfaces, or utility functions for each module:
  - **`kernels`**: Placeholder for CUDA kernel implementations.
  - **`models`**: Base classes for model architectures.
  - **`utils`**: Common utilities like logging, timers, and file I/O.
  - **`training`**: Training loops, loss functions, and evaluation metrics.
  - **`data`**: Data loaders, preprocessing pipelines, and dataset classes.
  - **`optim`**: Custom optimizers and learning rate schedulers.
  - **`config`**: Configuration management using YAML or JSON files.
  - **`scripts`**: Example scripts for training, inference, and profiling.

#### **4. Set Up Repository for Future Expansion**
- [ ] Ensure the structure is modular and scalable for adding new features or modules.
- [ ] Add a `README.md` file in each subdirectory to explain its purpose and usage.
- [ ] Include placeholder files (e.g., `example.py` or `template.py`) to guide contributors.

#### **5. Integrate with Existing Tools**
- [ ] Ensure compatibility with tools like **PyTorch**, **Hugging Face Transformers**, and **CUDA**.
- [ ] Add a `requirements.txt` or `setup.py` file to manage dependencies.

#### **6. Document the Structure**
- [ ] Update the main `README.md` to reflect the new directory structure.
- [ ] Provide a brief explanation of each module and its role in the project.

---

### **Context**
This task is part of **Phase 1** of the project roadmap. The goal is to establish a solid foundation for the repository, enabling future development of optimizations, training pipelines, and model implementations. For more details, refer to the [Project Board](https://github.com/orgs/eigencore/projects/1/views/1).

---

### **Suggested Modules and Their Purpose**

| **Module**   | **Purpose**                                                                 |
|--------------|-----------------------------------------------------------------------------|
| `kernels`    | Custom CUDA kernels for GPU-accelerated operations (e.g., matrix multiplication, attention mechanisms). |
| `models`     | Model architectures (e.g., LLaMA, transformers) and their components.       |
| `utils`      | Utility functions for logging, debugging, file handling, and other common tasks. |
| `training`   | Training scripts, loops, loss functions, and evaluation pipelines.          |
| `data`       | Data loading, preprocessing, and dataset management.                        |
| `optim`      | Custom optimizers, learning rate schedulers, and optimization algorithms.   |
| `config`     | Configuration management for hyperparameters and experiment settings.       |
| `scripts`    | Standalone scripts for tasks like profiling, benchmarking, and deployment.  |

---

### **Example Structure**
Here’s how the repository might look after implementing the structure:

```
tlama-core/
├── tlamacore/
│   ├── __init__.py
│   ├── kernels/
│   │   ├── __init__.py
│   │   ├── attention.py
│   │   └── matrix_ops.py
│   ├── models/
│   │   ├── __init__.py
│   │   ├── llama.py
│   │   └── transformer.py
│   ├── utils/
│   │   ├── __init__.py
│   │   ├── logger.py
│   │   └── file_io.py
│   ├── training/
│   │   ├── __init__.py
│   │   ├── trainer.py
│   │   └── metrics.py
│   ├── data/
│   │   ├── __init__.py
│   │   ├── dataloader.py
│   │   └── preprocessing.py
│   ├── optim/
│   │   ├── __init__.py
│   │   ├── optimizers.py
│   │   └── schedulers.py
│   ├── config/
│   │   ├── __init__.py
│   │   └── config_manager.py
│   └── scripts/
│       ├── __init__.py
│       ├── train.py
│       └── profile.py
├── requirements.txt
├── setup.py
└── README.md
```

---

### **Next Steps**
Once the basic infrastructure is in place, we can:
1. Implement custom CUDA kernels in the `kernels` module.
2. Add model architectures to the `models` module.
3. Develop training pipelines in the `training` module.
4. Expand the `data` module to support various datasets and preprocessing techniques.
